File: blendertools/SRM_Blender/dds.py
"""Minimal DDS texture decoder (stdlib only, no PIL).

Decodes the top mip of DXT1 / DXT3 / DXT5 (and simple uncompressed RGB/RGBA)
DirectDraw Surface files to a flat RGBA byte buffer, so textures can be loaded
into Blender's bundled Python which has no Pillow.

Usage:
    img = DDS.read(path)          # -> DDS with .width/.height/.rgba (bytes, R,G,B,A per px, top-down)
    rgba = img.rgba
"""

import struct

DDSD_MAGIC = b'DDS '

# ddspf.dwFlags
DDPF_FOURCC = 0x4
DDPF_RGB = 0x40
DDPF_ALPHAPIXELS = 0x1

FOURCC_DXT1 = b'DXT1'
FOURCC_DXT2 = b'DXT2'
FOURCC_DXT3 = b'DXT3'
FOURCC_DXT4 = b'DXT4'
FOURCC_DXT5 = b'DXT5'


class DDS:
    __slots__ = ('width', 'height', 'rgba')

    def __init__(self, width, height, rgba):
        self.width = width
        self.height = height
        self.rgba = rgba  # bytearray, len = width*height*4, row-major top-down

    @staticmethod
    def read(path):
        with open(path, 'rb') as f:
            data = f.read()
        return DDS.from_bytes(data)

    @staticmethod
    def from_bytes(data):
        if data[:4] != DDSD_MAGIC:
            raise ValueError("Not a DDS file")
        height = struct.unpack_from('<I', data, 12)[0]
        width = struct.unpack_from('<I', data, 16)[0]
        pf_flags = struct.unpack_from('<I', data, 80)[0]
        fourcc = data[84:88]
        rgb_bit_count = struct.unpack_from('<I', data, 88)[0]
        rmask, gmask, bmask, amask = struct.unpack_from('<IIII', data, 92)

        pixel_off = 128  # magic(4) + 124-byte header

        if pf_flags & DDPF_FOURCC:
            if fourcc == FOURCC_DXT1:
                rgba = _decode_dxt(data, pixel_off, width, height, dxt1=True)
            elif fourcc in (FOURCC_DXT2, FOURCC_DXT3):
                rgba = _decode_dxt3(data, pixel_off, width, height)
            elif fourcc in (FOURCC_DXT4, FOURCC_DXT5):
                rgba = _decode_dxt5(data, pixel_off, width, height)
            else:
                raise ValueError(f"Unsupported DDS fourCC {fourcc!r}")
        elif pf_flags & DDPF_RGB:
            rgba = _decode_uncompressed(data, pixel_off, width, height,
                                        rgb_bit_count, rmask, gmask, bmask,
                                        amask, pf_flags)
        else:
            raise ValueError("Unsupported DDS pixel format")

        return DDS(width, height, rgba)


# --- block helpers ---

def _rgb565(c):
    r = (c >> 11) & 0x1F
    g = (c >> 5) & 0x3F
    b = c & 0x1F
    return ((r << 3) | (r >> 2), (g << 2) | (g >> 4), (b << 3) | (b >> 2))


def _lerp(c0, c1, w0, w1):
    t = w0 + w1
    return ((c0[0] * w0 + c1[0] * w1) // t,
            (c0[1] * w0 + c1[1] * w1) // t,
            (c0[2] * w0 + c1[2] * w1) // t)


def _color_palette(data, off, punchthrough):
    """Return (palette[4] rgb, alpha[4]) for a DXT color block at off."""
    c0_raw, c1_raw = struct.unpack_from('<HH', data, off)
    c0 = _rgb565(c0_raw)
    c1 = _rgb565(c1_raw)
    if punchthrough and c0_raw <= c1_raw:
        pal = [c0, c1, _lerp(c0, c1, 1, 1), (0, 0, 0)]
        alpha = [255, 255, 255, 0]
    else:
        pal = [c0, c1, _lerp(c0, c1, 2, 1), _lerp(c0, c1, 1, 2)]
        alpha = [255, 255, 255, 255]
    return pal, alpha


def _blit(out, width, height, bx, by, pal, color_idx, alpha):
    """Write a decoded 4x4 color block into the RGBA output buffer."""
    for pi in range(16):
        px = bx * 4 + (pi & 3)
        py = by * 4 + (pi >> 2)
        if px >= width or py >= height:
            continue
        ci = (color_idx >> (2 * pi)) & 0x3
        r, g, b = pal[ci]
        o = (py * width + px) * 4
        out[o] = r
        out[o + 1] = g
        out[o + 2] = b
        out[o + 3] = alpha[pi]


def _decode_dxt(data, off, width, height, dxt1):
    out = bytearray(width * height * 4)
    bw = (width + 3) // 4
    bh = (height + 3) // 4
    for by in range(bh):
        for bx in range(bw):
            pal, a = _color_palette(data, off, punchthrough=dxt1)
            color_idx = struct.unpack_from('<I', data, off + 4)[0]
            alpha = [a[(color_idx >> (2 * pi)) & 0x3] for pi in range(16)]
            _blit(out, width, height, bx, by, pal, color_idx, alpha)
            off += 8
    return out


def _decode_dxt3(data, off, width, height):
    out = bytearray(width * height * 4)
    bw = (width + 3) // 4
    bh = (height + 3) // 4
    for by in range(bh):
        for bx in range(bw):
            alpha_bits = int.from_bytes(data[off:off + 8], 'little')
            alpha = [((alpha_bits >> (4 * pi)) & 0xF) * 17 for pi in range(16)]
            pal, _ = _color_palette(data, off + 8, punchthrough=False)
            color_idx = struct.unpack_from('<I', data, off + 12)[0]
            _blit(out, width, height, bx, by, pal, color_idx, alpha)
            off += 16
    return out


def _decode_dxt5(data, off, width, height):
    out = bytearray(width * height * 4)
    bw = (width + 3) // 4
    bh = (height + 3) // 4
    for by in range(bh):
        for bx in range(bw):
            a0 = data[off]
            a1 = data[off + 1]
            ap = [a0, a1]
            if a0 > a1:
                for i in range(2, 8):
                    ap.append((a0 * (8 - i) + a1 * (i - 1)) // 7)
            else:
                for i in range(2, 6):
                    ap.append((a0 * (6 - i) + a1 * (i - 1)) // 5)
                ap.append(0)
                ap.append(255)
            abits = int.from_bytes(data[off + 2:off + 8], 'little')
            alpha = [ap[(abits >> (3 * pi)) & 0x7] for pi in range(16)]
            pal, _ = _color_palette(data, off + 8, punchthrough=False)
            color_idx = struct.unpack_from('<I', data, off + 12)[0]
            _blit(out, width, height, bx, by, pal, color_idx, alpha)
            off += 16
    return out


def _mask_shift(mask):
    if mask == 0:
        return 0, 0
    shift = 0
    m = mask
    while not (m & 1):
        m >>= 1
        shift += 1
    bits = 0
    while m & 1:
        m >>= 1
        bits += 1
    return shift, bits


def _scale(val, bits):
    if bits == 0:
        return 255
    return (val * 255) // ((1 << bits) - 1)


def _decode_uncompressed(data, off, width, height, bpp, rmask, gmask, bmask,
                         amask, pf_flags):
    out = bytearray(width * height * 4)
    bytes_pp = bpp // 8
    rs, rb = _mask_shift(rmask)
    gs, gb = _mask_shift(gmask)
    bs, bb = _mask_shift(bmask)
    as_, ab = _mask_shift(amask)
    has_alpha = bool(pf_flags & DDPF_ALPHAPIXELS) and amask != 0
    for i in range(width * height):
        px = int.from_bytes(data[off:off + bytes_pp], 'little')
        off += bytes_pp
        o = i * 4
        out[o] = _scale((px & rmask) >> rs, rb)
        out[o + 1] = _scale((px & gmask) >> gs, gb)
        out[o + 2] = _scale((px & bmask) >> bs, bb)
        out[o + 3] = _scale((px & amask) >> as_, ab) if has_alpha else 255
    return out

File: viewer/src/dds.cpp
#include "dds.h"
#include <cstdio>
#include <cstring>

namespace {

struct RGB { int r, g, b; };

inline uint32_t rd32(const uint8_t* d, size_t off) {
    return (uint32_t)d[off] | ((uint32_t)d[off+1] << 8) |
           ((uint32_t)d[off+2] << 16) | ((uint32_t)d[off+3] << 24);
}
inline uint16_t rd16(const uint8_t* d, size_t off) {
    return (uint16_t)d[off] | ((uint16_t)d[off+1] << 8);
}
inline uint64_t rd64(const uint8_t* d, size_t off) {
    uint64_t v = 0;
    for (int i = 0; i < 8; i++) v |= (uint64_t)d[off + i] << (8 * i);
    return v;
}

RGB rgb565(uint16_t c) {
    int r = (c >> 11) & 0x1F, g = (c >> 5) & 0x3F, b = c & 0x1F;
    return { (r << 3) | (r >> 2), (g << 2) | (g >> 4), (b << 3) | (b >> 2) };
}
RGB lerp(RGB a, RGB b, int wa, int wb) {
    int t = wa + wb;
    return { (a.r*wa + b.r*wb)/t, (a.g*wa + b.g*wb)/t, (a.b*wa + b.b*wb)/t };
}

// Build the 4-entry color palette + implicit per-block alpha for a color block.
void colorPalette(const uint8_t* d, size_t off, bool punchthrough,
                  RGB pal[4], int blockAlpha[4]) {
    uint16_t c0raw = rd16(d, off), c1raw = rd16(d, off + 2);
    RGB c0 = rgb565(c0raw), c1 = rgb565(c1raw);
    if (punchthrough && c0raw <= c1raw) {
        pal[0]=c0; pal[1]=c1; pal[2]=lerp(c0,c1,1,1); pal[3]={0,0,0};
        blockAlpha[0]=255; blockAlpha[1]=255; blockAlpha[2]=255; blockAlpha[3]=0;
    } else {
        pal[0]=c0; pal[1]=c1; pal[2]=lerp(c0,c1,2,1); pal[3]=lerp(c0,c1,1,2);
        blockAlpha[0]=255; blockAlpha[1]=255; blockAlpha[2]=255; blockAlpha[3]=255;
    }
}

void blit(std::vector<uint8_t>& out, int W, int H, int bx, int by,
          const RGB pal[4], uint32_t colorIdx, const int alpha[16]) {
    for (int pi = 0; pi < 16; pi++) {
        int px = bx * 4 + (pi & 3);
        int py = by * 4 + (pi >> 2);
        if (px >= W || py >= H) continue;
        int ci = (colorIdx >> (2 * pi)) & 0x3;
        size_t o = (size_t)(py * W + px) * 4;
        out[o]   = (uint8_t)pal[ci].r;
        out[o+1] = (uint8_t)pal[ci].g;
        out[o+2] = (uint8_t)pal[ci].b;
        out[o+3] = (uint8_t)alpha[pi];
    }
}

std::vector<uint8_t> decodeDXT1(const uint8_t* d, size_t off, int W, int H) {
    std::vector<uint8_t> out((size_t)W * H * 4, 0);
    int bw = (W + 3) / 4, bh = (H + 3) / 4;
    for (int by = 0; by < bh; by++)
        for (int bx = 0; bx < bw; bx++) {
            RGB pal[4]; int ba[4];
            colorPalette(d, off, true, pal, ba);
            uint32_t colorIdx = rd32(d, off + 4);
            int alpha[16];
            for (int pi = 0; pi < 16; pi++) alpha[pi] = ba[(colorIdx >> (2*pi)) & 0x3];
            blit(out, W, H, bx, by, pal, colorIdx, alpha);
            off += 8;
        }
    return out;
}

std::vector<uint8_t> decodeDXT3(const uint8_t* d, size_t off, int W, int H) {
    std::vector<uint8_t> out((size_t)W * H * 4, 0);
    int bw = (W + 3) / 4, bh = (H + 3) / 4;
    for (int by = 0; by < bh; by++)
        for (int bx = 0; bx < bw; bx++) {
            uint64_t ab = rd64(d, off);
            int alpha[16];
            for (int pi = 0; pi < 16; pi++) alpha[pi] = (int)((ab >> (4*pi)) & 0xF) * 17;
            RGB pal[4]; int ba[4];
            colorPalette(d, off + 8, false, pal, ba);
            uint32_t colorIdx = rd32(d, off + 12);
            blit(out, W, H, bx, by, pal, colorIdx, alpha);
            off += 16;
        }
    return out;
}

std::vector<uint8_t> decodeDXT5(const uint8_t* d, size_t off, int W, int H) {
    std::vector<uint8_t> out((size_t)W * H * 4, 0);
    int bw = (W + 3) / 4, bh = (H + 3) / 4;
    for (int by = 0; by < bh; by++)
        for (int bx = 0; bx < bw; bx++) {
            int a0 = d[off], a1 = d[off + 1];
            int ap[8]; ap[0] = a0; ap[1] = a1;
            if (a0 > a1) for (int i = 2; i < 8; i++) ap[i] = (a0*(8-i) + a1*(i-1)) / 7;
            else { for (int i = 2; i < 6; i++) ap[i] = (a0*(6-i) + a1*(i-1)) / 5; ap[6]=0; ap[7]=255; }
            uint64_t abits = 0;
            for (int i = 0; i < 6; i++) abits |= (uint64_t)d[off + 2 + i] << (8 * i);
            int alpha[16];
            for (int pi = 0; pi < 16; pi++) alpha[pi] = ap[(abits >> (3*pi)) & 0x7];
            RGB pal[4]; int ba[4];
            colorPalette(d, off + 8, false, pal, ba);
            uint32_t colorIdx = rd32(d, off + 12);
            blit(out, W, H, bx, by, pal, colorIdx, alpha);
            off += 16;
        }
    return out;
}

void maskShift(uint32_t mask, int& shift, int& bits) {
    if (mask == 0) { shift = 0; bits = 0; return; }
    shift = 0; uint32_t m = mask;
    while (!(m & 1)) { m >>= 1; shift++; }
    bits = 0; while (m & 1) { m >>= 1; bits++; }
}
int scalec(uint32_t val, int bits) {
    if (bits == 0) return 255;
    return (int)((val * 255) / ((1u << bits) - 1));
}

std::vector<uint8_t> decodeUncompressed(const uint8_t* d, size_t off, int W, int H,
        int bpp, uint32_t rmask, uint32_t gmask, uint32_t bmask, uint32_t amask,
        bool hasAlpha) {
    std::vector<uint8_t> out((size_t)W * H * 4, 0);
    int bpp8 = bpp / 8;
    int rs, rb, gs, gb, bs, bb, as, ab;
    maskShift(rmask, rs, rb); maskShift(gmask, gs, gb);
    maskShift(bmask, bs, bb); maskShift(amask, as, ab);
    for (int i = 0; i < W * H; i++) {
        uint32_t px = 0;
        for (int k = 0; k < bpp8; k++) px |= (uint32_t)d[off + k] << (8 * k);
        off += bpp8;
        size_t o = (size_t)i * 4;
        out[o]   = (uint8_t)scalec((px & rmask) >> rs, rb);
        out[o+1] = (uint8_t)scalec((px & gmask) >> gs, gb);
        out[o+2] = (uint8_t)scalec((px & bmask) >> bs, bb);
        out[o+3] = hasAlpha ? (uint8_t)scalec((px & amask) >> as, ab) : 255;
    }
    return out;
}

} // namespace

DdsImage dds_load(const std::string& path) {
    DdsImage img;
    FILE* f = fopen(path.c_str(), "rb");
    if (!f) return img;
    fseek(f, 0, SEEK_END); long sz = ftell(f); fseek(f, 0, SEEK_SET);
    if (sz < 128) { fclose(f); return img; }
    std::vector<uint8_t> data((size_t)sz);
    if (fread(data.data(), 1, (size_t)sz, f) != (size_t)sz) { fclose(f); return img; }
    fclose(f);

    if (memcmp(data.data(), "DDS ", 4) != 0) return img;
    const uint8_t* d = data.data();
    int H = (int)rd32(d, 12), W = (int)rd32(d, 16);
    uint32_t pfFlags = rd32(d, 80);
    char fourcc[5] = {0}; memcpy(fourcc, d + 84, 4);
    int rgbBits = (int)rd32(d, 88);
    uint32_t rmask = rd32(d, 92), gmask = rd32(d, 96), bmask = rd32(d, 100), amask = rd32(d, 104);
    const size_t pixel = 128;

    const uint32_t DDPF_FOURCC = 0x4, DDPF_RGB = 0x40, DDPF_ALPHAPIXELS = 0x1;
    if (pfFlags & DDPF_FOURCC) {
        if (!memcmp(fourcc, "DXT1", 4)) img.rgba = decodeDXT1(d, pixel, W, H);
        else if (!memcmp(fourcc, "DXT2", 4) || !memcmp(fourcc, "DXT3", 4)) img.rgba = decodeDXT3(d, pixel, W, H);
        else if (!memcmp(fourcc, "DXT4", 4) || !memcmp(fourcc, "DXT5", 4)) img.rgba = decodeDXT5(d, pixel, W, H);
        else return img;
    } else if (pfFlags & DDPF_RGB) {
        bool hasAlpha = (pfFlags & DDPF_ALPHAPIXELS) && amask != 0;
        img.rgba = decodeUncompressed(d, pixel, W, H, rgbBits, rmask, gmask, bmask, amask, hasAlpha);
    } else {
        return img;
    }
    img.width = W; img.height = H; img.ok = true;
    return img;
}
